extractdilkey.c

```diff
@@ -0,0 +1,386 @@
+                sRc = writeFile(sRawKey, sRawBytes, sOutFile);
+            }
+            else
+            {
+                sRc = writeFile(sKey, sKeyBytes, sOutFile);
+            }
+        }
+        else
+        {
+            printf("Valid Dilithium public keyfile detected\n");
+        }
+    }
+
+    // Private keys
+    else if (0 == sRc && !sPubIn)
+    {
+        if (sRawIn)
+        {
+            // Raw private key size for dilithium r2 8/7
+            if (5136 == sKeyBytes)
+            {
+                if (sVerbose) printf("extractdilkey: Found raw private key\n");
+                // We have a raw private key, lets convert it
+                memcpy(sRawKey, sKey, sKeyBytes);
+                sRawBytes = sKeyBytes;
+
+                sKeyBytes = BUF_SIZE;
+
+                // TODO , convert raw private key to encoded format without the public key
+                if (sOutFile)
+                {
+                    sRc = 1;
+                    printf("**** ERROR: Unable to convert private raw -> encoded\n");
+                    break;
+                }
+            }
+            else
+            {
+                printf("**** ERROR: Unrecognized raw private key : %s\n", sInFile);
+                sRc = 1;
+                break;
+            }
+        }
+        else
+        {
+            // Attempt to convert encoded key
+            unsigned int sWireType = 0;
+            if (sPubOut)
+            {
+                // Get the raw public key
+                sRc = mlca_wire2key(sRawKey, sRawBytes, &sWireType,
+                                    sKey, sKeyBytes,
+                                    (const unsigned char *) CR_OID_SPECIAL_PRV2PUB,
+                                    CR_OID_SPECIAL_PRV2PUB_BYTES);
+                if (0 >= sRc)
+                {
+                    printf("**** ERROR: Unable to convert private key : %d\n",sRc);
+                    sRc = 1;
+                    break;
+                }
+                else
+                {
+                    sRawBytes = sRc;
+                    sRc = 0;
+                }
+                if (sVerbose) printf("extractdilkey: Found public key\n");
+
+                // Encode it
+                sRc = mlca_key2wire(sKey,    sKeyBytes,
+                                    sRawKey, sRawBytes, 0,
+                                    NULL, 0,
+                                    NULL, 0);
+                if (sRc < 0)
+                {
+                    printf("**** ERROR: Failure during public key conversion : %d\n", sRc);
+                    break;
+                }
+                sKeyBytes = sRc;
+            }
+            else
+            {
+                sRc = mlca_wire2key(sRawKey, sRawBytes, &sWireType,
+                                    sKey, sKeyBytes,
+                                    NULL, ~0);
+
+                // Raw private key size for dilithium r2 8/7
+                if (0 >= sRc || 5136 != sRc)
```

Comment:
Now 0 and positive rcs are failures?

Reply:
As said above the return code is the size of the key.  So this checks if negative or if the size is not as expected treat as failure